Source/WebCore/html/ValidatedFormListedElement.cpp

```diff
@@ -0,0 +1,495 @@
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ValidatedFormListedElement.h"
+
+#include "ControlStates.h"
+#include "ElementAncestorIterator.h"
+#include "Event.h"
+#include "EventHandler.h"
+#include "EventNames.h"
+#include "FormAssociatedElement.h"
+#include "FormController.h"
+#include "Frame.h"
+#include "FrameView.h"
+#include "HTMLDataListElement.h"
+#include "HTMLFieldSetElement.h"
+#include "HTMLFormElement.h"
+#include "HTMLLegendElement.h"
+#include "HTMLParserIdioms.h"
+#include "PseudoClassChangeInvalidation.h"
+#include "ScriptDisallowedScope.h"
+#include "ValidationMessage.h"
+#include <wtf/Ref.h>
+#include <wtf/SetForScope.h>
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+using namespace HTMLNames;
+
+ValidatedFormListedElement::ValidatedFormListedElement(HTMLFormElement* form)
+    : FormListedElement { form }
+{
+}
+
+ValidatedFormListedElement::~ValidatedFormListedElement()
+{
+    clearForm();
+}
+
+bool ValidatedFormListedElement::willValidate() const
+{
+    if (!m_willValidateInitialized || m_isInsideDataList == TriState::Indeterminate) {
+        m_willValidateInitialized = true;
+        bool newWillValidate = computeWillValidate();
+        if (m_willValidate != newWillValidate)
+            m_willValidate = newWillValidate;
+    } else {
+        // If the following assertion fails, updateWillValidateAndValidity() is not
+        // called correctly when something which changes computeWillValidate() result
+        // is updated.
+        ASSERT(m_willValidate == computeWillValidate());
+    }
+    return m_willValidate;
+}
+
+bool ValidatedFormListedElement::computeWillValidate() const
+{
+    if (m_isInsideDataList == TriState::Indeterminate) {
+#if ENABLE(DATALIST_ELEMENT)
+        const HTMLElement& element = asHTMLElement();
+        m_isInsideDataList = triState(element.document().hasDataListElements() && ancestorsOfType<HTMLDataListElement>(element).first());
+#else
+        m_isInsideDataList = TriState::False;
+#endif
+    }
+    // readonly bars constraint validation for *all* <input> elements, regardless of the <input> type, for compat reasons.
+    return m_isInsideDataList == TriState::False && !isDisabled() && !m_hasReadOnlyAttribute;
+}
+
+void ValidatedFormListedElement::updateVisibleValidationMessage(Ref<HTMLElement> anchor)
```

Comment:
I think it's very confusing to call this function argument "anchor" given we have HTMLAnchorElement.
Why not spell out the full name: validationAnchor.

Reply:
Renamed.